File: android/app/src/main/java/io/github/magnusencoded/stationtostation/data/gossip/GossipTally.kt
package io.github.magnusencoded.stationtostation.data.gossip

import java.util.concurrent.atomic.AtomicLong

/**
 * Six numbers that say whether routing receipts did anything tonight (#444).
 *
 * ADR-0022 ships receipts and the ranking that reads them knowing that several independent
 * things could make the whole feature a no-op in the field, and that none of them is visible
 * from the outside: a receipt that is never authored, one that is never owed because the
 * **Pass** could not be addressed, one that is authored but never offered,
 * one that is offered but never delivered, and a ranking that never finds live credit for a
 * peer it can see all present identically — as "no preference", which is also what correct
 * looks like when nobody is around. The point of these counters is to tell those apart at the
 * end of a night with one line instead of a night of logcat.
 *
 * **Cheap and off the hot path by construction.** Six `AtomicLong`s incremented at a handful of places,
 * no allocation, no flow, no persistence. It follows [GossipRadioStatus]'s rule — diagnostic,
 * process-wide, losing everything to a process death is correct rather than a gap — with the
 * one difference that [GossipRadioStatus.radioStopped] wipes its state and this deliberately
 * does not: a tally that resets when the relay stops is a tally you can never read, because
 * stopping is when you go looking.
 *
 * These are counts of events, never of peers, and nothing here names anybody. A number is not
 * a routing fact about a person, which is the whole reason it is safe to log.
 */
object GossipTally {

    private val authored = AtomicLong()
    private val declined = AtomicLong()
    private val offered = AtomicLong()
    private val delivered = AtomicLong()
    private val windows = AtomicLong()
    private val creditHits = AtomicLong()

    /** A receipt this device wrote, on admitting a **Fact** it recognised as a **Contact**'s. */
    fun authored(count: Int = 1) { if (count > 0) authored.addAndGet(count.toLong()) }

    /**
     * A **Pass** this device owed nothing on because it could not address its sender
     * (ADR-0022 §2a: the **Pass** was signed as a **Gig**, carrying its signer's own request).
     *
     * One per **Pass**, not per **Fact**, because the decision is about the **Pass**. This is
     * the price of the addressing fix made visible: §2a argues the credit is deferred to that
     * neighbour's next **Pass** rather than lost, and the ratio of this to [authored] is what
     * decides whether that argument survives contact with a real room.
     */
    fun declined() { declined.incrementAndGet() }

    /** Receipts put into a **Pass** — offered is not delivered, and the gap is the interesting part. */
    fun offered(count: Int) { if (count > 0) offered.addAndGet(count.toLong()) }

    /** Receipts in a **Pass** the peer acknowledged to the last chunk. */
    fun delivered(count: Int) { if (count > 0) delivered.addAndGet(count.toLong()) }

    /**
     * One closed pick window: how many candidates it ranked, and how many of those the
     * ranker could find live credit for.
     *
     * [hits] is the number ADR-0022 calls the cheapest falsifier. If it stays at zero across a
     * whole night with a non-zero [windows], `gossipPreferredPeers` is a shuffle and nothing
     * more, and no amount of tuning the decay changes that.
     */
    fun ranked(candidates: Int, hits: Int) {
        if (candidates <= 0) return
        windows.incrementAndGet()
        if (hits > 0) creditHits.addAndGet(hits.toLong())
    }

    /** The whole night in one line, for a log at shutdown or a panel someone is staring at. */
    fun summary(): String = "receipts authored=${authored.get()} declined=${declined.get()} " +
        "offered=${offered.get()} delivered=${delivered.get()} · " +
        "pick windows=${windows.get()} credit hits=${creditHits.get()}"

    /** Only for tests; a running relay never resets its own tally. */
    internal fun reset() {
        listOf(authored, declined, offered, delivered, windows, creditHits).forEach { it.set(0) }
    }
}

File: ios/StationToStation/Data/Gossip/GossipTally.swift
import Foundation

/// Six numbers that say whether routing receipts did anything tonight (#444).
///
/// The twin of Android's `GossipTally`, and for the same reason: ADR-0022 ships receipts
/// knowing that several independent things could make the whole feature a no-op in the field,
/// and that all of them look identical from the outside — "no preference", which is also what
/// correct looks like when nobody is around. A receipt never authored, one never owed because
/// the **Pass** could not be addressed (ADR-0022 §2a), one authored but never offered, one
/// offered but never delivered, and a ranking that never finds live credit are five different
/// failures with one symptom, and this is what tells them apart.
///
/// **Cheap and off the hot path by construction.** Six counters behind one lock, incremented
/// at a handful of places, nothing persisted. Losing it all to a process death is correct
/// rather than a gap.
///
/// `creditHits` is live on both platforms as of #486: `GossipTransport` gathers a pick window
/// and ranks it against `gossipMaxConcurrentMeetings` free slots, so a window that finds no
/// credit is now a measurement rather than a missing caller. A zero here means the ranking ran
/// and found nothing — which, given that credit is only learnable after a peer has been met
/// once (ADR-0022 §4), is still the expected reading early in a night.
///
/// These are counts of events, never of peers, and nothing here names anybody.
final class GossipTally: @unchecked Sendable {
    static let shared = GossipTally()

    private let lock = NSLock()
    private var authoredCount: Int64 = 0
    private var declinedCount: Int64 = 0
    private var offeredCount: Int64 = 0
    private var deliveredCount: Int64 = 0
    private var windowCount: Int64 = 0
    private var creditHitCount: Int64 = 0

    /// Receipts this device wrote, on admitting **Facts** it recognised as a **Contact**'s.
    func authored(_ count: Int = 1) { bump(count) { $0.authoredCount += $1 } }

    /// A **Pass** this device owed nothing on because it could not address its sender
    /// (ADR-0022 §2a: signed as a **Gig**, because it carried its signer's own request).
    ///
    /// One per **Pass**, not per **Fact**. The ratio of this to `authored` is what decides
    /// whether §2a's "deferred, not lost" argument survives a real room.
    func declined() { bump(1) { $0.declinedCount += $1 } }

    /// Receipts put into a **Pass** — offered is not delivered, and the gap is the interesting part.
    func offered(_ count: Int) { bump(count) { $0.offeredCount += $1 } }

    /// Receipts in a **Pass** the peer took to the end.
    func delivered(_ count: Int) { bump(count) { $0.deliveredCount += $1 } }

    /// One closed pick window: how many candidates it ranked, and how many held live credit.
    func ranked(candidates: Int, hits: Int) {
        guard candidates > 0 else { return }
        lock.lock(); defer { lock.unlock() }
        windowCount += 1
        if hits > 0 { creditHitCount += Int64(hits) }
    }

    /// The whole night in one line, for a log at shutdown.
    func summary() -> String {
        lock.lock(); defer { lock.unlock() }
        return "receipts authored=\(authoredCount) declined=\(declinedCount) "
            + "offered=\(offeredCount) delivered=\(deliveredCount) · "
            + "pick windows=\(windowCount) credit hits=\(creditHitCount)"
    }

    /// Only for tests; a running relay never resets its own tally.
    func reset() {
        lock.lock(); defer { lock.unlock() }
        authoredCount = 0; declinedCount = 0; offeredCount = 0; deliveredCount = 0
        windowCount = 0; creditHitCount = 0
    }

    private func bump(_ count: Int, _ apply: (GossipTally, Int64) -> Void) {
        guard count > 0 else { return }
        lock.lock(); defer { lock.unlock() }
        apply(self, Int64(count))
    }
}
